pkg/scheduler/util/non_zero.go

```diff
@@ -34,45 +37,72 @@ const (
-	return GetNonzeroRequestForResource(v1.ResourceCPU, requests),
-		GetNonzeroRequestForResource(v1.ResourceMemory, requests)
+func GetNonzeroRequests(pod *v1.Pod) (int64, int64) {
+	return GetNonzeroRequestForResource(pod, v1.ResourceCPU),
+		GetNonzeroRequestForResource(pod, v1.ResourceMemory)
 }
 
 // GetNonzeroRequestForResource returns the default resource request if none is found or
 // what is provided on the request.
-func GetNonzeroRequestForResource(resource v1.ResourceName, requests *v1.ResourceList) int64 {
-	switch resource {
-	case v1.ResourceCPU:
-		// Override if un-set, but not if explicitly set to zero
-		if _, found := (*requests)[v1.ResourceCPU]; !found {
-			return DefaultMilliCPURequest
-		}
-		return requests.Cpu().MilliValue()
-	case v1.ResourceMemory:
-		// Override if un-set, but not if explicitly set to zero
-		if _, found := (*requests)[v1.ResourceMemory]; !found {
-			return DefaultMemoryRequest
+func GetNonzeroRequestForResource(pod *v1.Pod, resource v1.ResourceName) int64 {
+	podRequest := resourcehelper.GetResourceRequest(pod, resource)
+	resourceFound := true
+
+	// If Overhead is being utilized, add to the total requests for the pod
+	for i := range pod.Spec.Containers {
+		container := &pod.Spec.Containers[i]
+		requests := container.Resources.Requests
+
+		// If the pod has an explicit resource request(includes initContainers),
+		// we shouldn't assign a default value to it.
+		if podRequest != 0 {
+			break
 		}
-		return requests.Memory().Value()
-	case v1.ResourceEphemeralStorage:
-		quantity, found := (*requests)[v1.ResourceEphemeralStorage]
-		if !found {
-			return 0
+
+		switch resource {
+		case v1.ResourceCPU:
+			// Override if un-set, but not if explicitly set to zero
+			if _, found := (requests)[v1.ResourceCPU]; !found {
+				podRequest = DefaultMilliCPURequest
+				resourceFound = false
+			}
+		case v1.ResourceMemory:
+			// Override if un-set, but not if explicitly set to zero
+			if _, found := (requests)[v1.ResourceMemory]; !found {
+				podRequest = DefaultMemoryRequest
+				resourceFound = false
+			}
+		case v1.ResourceEphemeralStorage:
+			if _, found := (requests)[v1.ResourceEphemeralStorage]; !found {
+				podRequest = DefaultEphemeralStorage
+				resourceFound = false
+			}
+		default:
+			if v1helper.IsScalarResourceName(resource) {
+				quantity, found := (requests)[resource]
+				if !found {
+					podRequest = 0
+					resourceFound = false
+				}
+				podRequest = quantity.Value()
+			}
 		}
-		return quantity.Value()
-	default:
-		if v1helper.IsScalarResourceName(resource) {
-			quantity, found := (*requests)[resource]
-			if !found {
-				return 0
+	}
+
+	if pod.Spec.Overhead != nil && utilfeature.DefaultFeatureGate.Enabled(features.PodOverhead) {
+		if quantity, found := pod.Spec.Overhead[resource]; found && !resourceFound {
+			if resource == v1.ResourceCPU {
+				podRequest += quantity.MilliValue()
+			} else {
+				podRequest += quantity.Value()
```

Comment:
Please pay attention on the condition for the if statement
```
if quantity, found := pod.Spec.Overhead[resource]; found && !resourceFound 
```
If we only specify the memory.Overhead value, but the memory.request is 0, the `resourcehelper.GetResourceRequest` will ignore the overhead value and return 0. So I just add the resourceFound variable, if the request wasn't specified and the overhead value is not nil, we need to calculate it by ourself.

Reply:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/resource/helpers.go#L123
Pay attention on the condition `!requestQuantity.IsZero()`, if we didn't assign the resource request, it will ignore the overhead value

```
	// if PodOverhead feature is supported, add overhead for running a pod
	// to the total requests if the resource total is non-zero
	if pod.Spec.Overhead != nil && utilfeature.DefaultFeatureGate.Enabled(features.PodOverhead) {
		if podOverhead, ok := pod.Spec.Overhead[resourceName]; ok && !requestQuantity.IsZero() {
			requestQuantity.Add(podOverhead)
		}
	}

	return requestQuantity
}
```

Reply:
Our original implementation to handle the overHead value in the node_info.go file is different from resourceHelper
```
// Old implementation, we didn't care about if the cpu request or memory request is zero.

		if _, found := pod.Spec.Overhead[v1.ResourceCPU]; found {	
			non0CPU += pod.Spec.Overhead.Cpu().MilliValue()	
		}	

		if _, found := pod.Spec.Overhead[v1.ResourceMemory]; found {	
			non0Mem += pod.Spec.Overhead.Memory().Value()	
		}
```

Reply:
Like aforementioned, let's make `resourcehelper.GetResourceRequest` also fit scheduling's purpose, so we don't need to calculate overhead again.
In Scoring's perspective, the goal is to give a non-zero value for BestEfforts Pod - so I'd suggest to use `max(max(initContainers),sum(containers)) + sum(overhead) !=0 ? <that value> : default value`.
BTW: if we follow this, we need to update some tests, as before we summed the value `(container req != 0 ? container req : default value) + overhead` for each container.